### PR TITLE
New test helpers

### PR DESCRIPTION
**Pull request by deltics**
## Release Notes

- **New Features**
  - Enhanced documentation in the README with sections on Basic and Advanced Usage for improved user guidance.
  - Introduced new testing utilities for context handling, including `ContextIndicator` and `ContextValue`.
  - Added a generic mocking framework with `MockFn` for flexible function testing.
  - New `Returns` method in the `Fake` type for setting result and error values.

- **Bug Fixes**
  - Improved error handling with new error variables for better reporting.

- **Tests**
  - Comprehensive test coverage added for new features, including tests for `MockFn` and context-related functionality.
  - Expanded test cases for the `Fake` type to validate various return scenarios.

- **Documentation**
  - Improved readability and navigation in the README and other documentation files, including clearer guidance on using the `Fake` type and mocking framework.

### CODE REVIEW

File: context_test.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+type key int
+
+var ctxKey key = 1
+
+func ContextKey(ctx context.Context) (string, bool) {
+	if v := ctx.Value(ctxKey); v != nil {
+		return v.(string), true
+	}
+	return "", false
+}
```
[the remark this answers is not in the document]
Comment (deltics) [reply context_test.go:17]: this function is a test fixture, currently only used with test values known to be of type string.  This will be addressed if necessary in the future if it becomes necessary to avoid such panics as the result of additional test cases.

File: mockFn.go [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+// fails immediately without yielding potentially incorrect results.
+func (mock *mockFnCall[A, R]) WillReturn(v ...any) {
+	resultSet := false
+	errSet := false
+	for _, r := range v {
+		switch r := r.(type) {
+		case R:
+			if resultSet {
+				panic(fmt.Errorf("%w: only one error value may be configured", ErrInvalidOperation))
+			}
```
[the remark this answers is not in the document]
Comment (deltics) [reply mockFn.go:164]: Thank you, will be fixed!

File: mockFn.go [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+//		MyMethod(int) (int, error)
+//	}
+//
+//	type myMock struct {
+//		myMethod test.MockFn[int, int]
+//	}
+//
+//	func (mock *myMock) MyMethod(arg int) (int, error) {
+//		return mock.myMethod.CalledWith(arg)
+//	}
```
[the remark this answers is not in the document]
Comment (deltics) [reply mockFn.go:223]: Thank you, will be fixed!

File: mockFn.go [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+	}
+
+	err = mock.expected.err
+	returns = mock.expected.result
+
+	if mock.expected.args != nil && *mock.expected.args != *actual.args {
+		actual.err = ErrUnexpectedArgs
+		err = fmt.Errorf("%w:\n  expected: %v\n  got     : %v", ErrUnexpectedArgs, mock.expected.args, actual.args)
+		mock.errs = append(mock.errs, err)
+	}
```
[the remark this answers is not in the document]
Comment (deltics) [reply mockFn.go:262]: I have refactored the conditions for testing expected vs actual args to cover this and other scenarios.

File: mockFn.go [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+//
+// When mocking a function that returns multiple result values (in addition to an error),
+// use a struct type to specify the return type, with fields for each of the result values.
+//
+// # Testing and Recording Arguments (Spy)
+//
+// MockFn does not implement the function being mocked; this still requires a mock
+// implementation to provide that function, whichA MockFn can be used to spy on the arguments passed to a function
+// It is useful for creating a spy to record the arguments and provide fake
+// return values of a function call.
```
[the remark this answers is not in the document]
Comment (deltics) [reply mockFn.go:55]: Fixed

File: fake_test.go [hunk 6]
```diff
@@ -12,3 +15,86 @@ func TestFake(t *testing.T) {
...
+				sut := Fake[int]{}
+
+				// ACT
+				sut.Returns(42)
+
+				// ASSERT
+				That(t, sut).Equals(Fake[int]{Result: 42})
+			},
+		},
+		{scenario: "returns error}",
```
[the remark this answers is not in the document]
Comment (deltics) [reply fake_test.go:37]: Thanks!

File: mockFn_test.go [hunk 7]
```diff
@@ -0,0 +1,463 @@
...
+				That(t, sut.expected).Equals(result)
+			},
+		},
+	}
+	for _, tc := range testcases {
+		t.Run(tc.scenario, func(t *testing.T) {
+			tc.exec(t)
+		})
+	}
+}
```
[the remark this answers is not in the document]
Comment (deltics) [reply mockFn_test.go:232]: Additional test case added to cover multiple calls and call chaining.

Calling ExpectCall() after a Reset() is not an edge case and does not require coverage; Reset() restores the MockFn to it's initial (zero) state (covered by a separate test) and a subsequent call to ExpectCall() is functionally equivalent to the first call to ExpectCall() on a new MockFn(), which is already covered.

File: mockFn_test.go [hunk 8]
```diff
@@ -0,0 +1,463 @@
...
+				sut.WillReturn(err, err)
+			},
+		},
+	}
+	for _, tc := range testcases {
+		t.Run(tc.scenario, func(t *testing.T) {
+			tc.exec(t)
+		})
+	}
+}
```
[the remark this answers is not in the document]
Comment (deltics) [reply mockFn_test.go]: Additional test case added.

File: mockFn.go [hunk 9]
```diff
@@ -0,0 +1,406 @@
...
+	mock.expected = nil
+
+	if mock.idxExpected < len(mock.expectations) {
+		mock.expected = mock.expectations[mock.idxExpected]
+	} else {
+		mock.idxExpected = -1
+	}
+
+	return
+}
```
[the remark this answers is not in the document]
Comment (deltics) [reply mockFn.go:283]: Making mock function thread-safe would be inappropriate; mocks are not intended to be used in concurrent tests since without predictable serialisation of calls it would be impossible to configure expected calls correctly.

When running concurrent tests, the test author must simply take care to ensure that each test uses a separate and distinct mock.
